src/coreclr/jit/fgprofilesynthesis.cpp

```diff
@@ -0,0 +1,917 @@
+            break;
+    }
+
+    // Determine cyclic probabilities
+    //
+    ComputeCyclicProbabilities();
+
+    // Assign weights to entry points in the flow graph
+    //
+    AssignInputWeights();
+
+    // Compute the block weights given the inputs and edge likelihoods
+    //
+    ComputeBlockWeights();
+}
+
+//------------------------------------------------------------------------
+// AssignLikelihoods: update edge likelihoods and block counts based
+//   entrely on heuristics.
+//
+// Notes:
+//   any existing likelihoods are removed in favor of heuristic
+//   likelihoods
+//
+void ProfileSynthesis::AssignLikelihoods()
+{
+    JITDUMP("Assigning edge likelihoods based on heuristics\n");
+
+    for (BasicBlock* const block : m_comp->Blocks())
+    {
+        switch (block->bbJumpKind)
+        {
+            case BBJ_THROW:
+            case BBJ_RETURN:
+            case BBJ_EHFINALLYRET:
+                // No successor cases
+                // (todo: finally ret may have succs)
+                break;
+
+            case BBJ_NONE:
+            case BBJ_CALLFINALLY:
+                // Single successor next cases
+                //
+                // Note we handle flow to the finally
+                // specially; this represents return
+                // from the finally.
+                AssignLikelihoodNext(block);
+                break;
+
+            case BBJ_ALWAYS:
+            case BBJ_LEAVE:
+            case BBJ_EHCATCHRET:
+            case BBJ_EHFILTERRET:
+                // Single successor jump cases
+                AssignLikelihoodJump(block);
+                break;
+
+            case BBJ_COND:
+                // Two successor cases
+                AssignLikelihoodCond(block);
+                break;
+
+            case BBJ_SWITCH:
+                // N successor cases
+                AssignLikelihoodSwitch(block);
+                break;
+
+            default:
+                unreached();
+        }
+    }
+}
+
+//------------------------------------------------------------------------
+// IsDfsAncestor: see if block `x` is ancestor of block `y` in the depth
+//   first spanning tree
+//
+// Arguments:
+//   x -- block that is possible ancestor
+//   y -- block that is possible descendant
+//
+// Returns:
+//   True if x is ancestor of y in the depth first spanning tree.
+//
+// Notes:
+//   If return value is false, then x does not dominate y.
+//
+bool ProfileSynthesis::IsDfsAncestor(BasicBlock* x, BasicBlock* y)
+{
+    return ((x->bbPreorderNum <= y->bbPreorderNum) && (y->bbPostorderNum <= x->bbPostorderNum));
+}
+
+//------------------------------------------------------------------------
+// IsLoopHeader: see if a block is a loop header, and if so return
+//   the associated loop.
+//
+// Arguments:
+//    block - block in question
+//
+// Returns:
+//    loop headed by block, or nullptr
+//
+SimpleLoop* ProfileSynthesis::IsLoopHeader(BasicBlock* block)
```

Comment:
`Is` implies a `bool` return. Maybe `GetLoopFromHeaderBlock`?

Reply:
Sure